### ISSUE TITLE
Forward test events to listener as they're emitted

### ISSUE THREAD

**Issue by Duhemm** [sbt/sbt#7435]
## Current situation
sbt provides an API to register test event listeners which help better understand test execution. At the moment, it's impossible to rely on the timing of these events, because:
 - they're not passed on to the listeners at the time they're emitted
 - `startGroup` and `endGroup` events are received by listeners at different times depending on whether the tests run in-process or in a forked JVM.

## What I'd like to do
I would like to fix this by:
 - changing the in-process test runner so that events are immediately sent to the listeners as they're produced by the test frameworks.
 - slightly change the communication protocol between sbt and the forked JVM:
   - introduce new `StartGroup` and `EndGroup` messages, which sbt will map to the eponymous events of the test listeners.
   - send test events immediately as they're produced by the test frameworks.
   - Increase the priority of the thread reading messages from the forked JVM, so that they're handled as early as possible.

No API changes are envisioned for this work.

## Benefits
This change will make it possible to measure how much time is spent doing test class setup and cleanup. Currently, it's not possible to get this information, because the `startGroup`, `testEvent` and `endGroup` callbacks are called together and their payloads don't contain this information.

## Limitations
The test events have no timestamp, so it's impossible for test events listeners to know accurately when an event was emitted. The proposed changes improves the situation, but the results cannot be 100% accurate. It'd be good to look into this for sbt 2.

### When forking
 - All events are collected for a given test group: https://github.com/sbt/sbt/blob/fb8ca26e77d495cf206a45025a63c9a6d0754a9f/testing/agent/src/main/java/sbt/ForkMain.java#L407-L412
 - The events are sent back together to sbt: https://github.com/sbt/sbt/blob/fb8ca26e77d495cf206a45025a63c9a6d0754a9f/testing/agent/src/main/java/sbt/ForkMain.java#L438
 - sbt reads the events and emits the `startGroup`, `testEvent` and `endGroup` events all together: https://github.com/sbt/sbt/blob/fb8ca26e77d495cf206a45025a63c9a6d0754a9f/main-actions/src/main/scala/sbt/ForkTests.scala#L215-L221

### When running in-process
 - The start and end group events are produced as early as possible: https://github.com/sbt/sbt/blob/fb8ca26e77d495cf206a45025a63c9a6d0754a9f/testing/src/main/scala/sbt/TestFramework.scala#L166-L170
 - The test events are collected and sent together when a test group finishes executing: https://github.com/sbt/sbt/blob/fb8ca26e77d495cf206a45025a63c9a6d0754a9f/testing/src/main/scala/sbt/TestFramework.scala#L137C14-L137C14

**Comment by claytonlin1110** [comment]
@eed3si9n Can I work on this issue?

**Comment by eed3si9n** [comment]
@claytonlin1110 Yes, please. The situation around forked test may have changed a bit in sbt 2.x so the locations of some of the linked sources may have shifted, so pleaes make sure that the problem still exists in sbt 2.x (freshly baked develop branch) and/or 2.0.0-RC12.

Please check out our [Contributor's Guildeline](https://github.com/sbt/sbt/blob/develop/CONTRIBUTING.md) and:
1. Make sure you can reproduce the problem and the fix manually (we regularly block folks who send untested or speculative PRs)
2. Include tests for the change
3. Follow our AI assisted contributions guideline
4. Multiple contributors might work on this issue in parallel
5. Generally we recommend getting CI to pass on a forked repo before sending a pull request
6. Sign Scala CLA <https://contribute.akka.io/contribute/cla/scala>
